[SPARK-18874][SQL] First phase: Deferring the correlated predicate pull up to Optimizer phase

dilipbiswal opened the pull request:
## What changes were proposed in this pull request?
Currently Analyzer as part of ResolveSubquery, pulls up the correlated predicates to its
originating SubqueryExpression. The subquery plan is then transformed to remove the correlated
predicates after they are moved up to the outer plan. In this PR, the task of pulling up
correlated predicates is deferred to Optimizer. This is the initial work that will allow us to
support the form of correlated subqueries that we don't support today. The design document
from @nsyca can be found in the following link :
[DesignDoc](https://docs.google.com/document/d/1QDZ8JwU63RwGFS6KVF54Rjj9ZJyK33d49ZWbjFBaIgU/edit#)

The brief description of code changes (hopefully to aid with code review) can be be found in the
following link:
[CodeChanges](https://docs.google.com/document/d/18mqjhL9V1An-tNta7aVE13HkALRZ5GZ24AATA-Vqqf0/edit#)


## How was this patch tested?
The test case PRs were submitted earlier using.
[16337](https://github.com/apache/spark/pull/16337) [16759](https://github.com/apache/spark/pull/16759) [16841](https://github.com/apache/spark/pull/16841) [16915](https://github.com/apache/spark/pull/16915) [16798](https://github.com/apache/spark/pull/16798) [16712](https://github.com/apache/spark/pull/16712) [16710](https://github.com/apache/spark/pull/16710) [16760](https://github.com/apache/spark/pull/16760) [16802](https://github.com/apache/spark/pull/16802)

dilipbiswal committed:
[SPARK-18874] First phase: Deferring the correlated predicate pull up to Optimizer phase

SparkQA commented:
**[Test build #72992 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72992/testReport)** for PR 16954 at commit [`33844bb`](https://github.com/apache/spark/commit/33844bb8c6e72c35e12a7ea9f124ad3c6b8bd43c).

SparkQA commented:
**[Test build #72992 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72992/testReport)** for PR 16954 at commit [`33844bb`](https://github.com/apache/spark/commit/33844bb8c6e72c35e12a7ea9f124ad3c6b8bd43c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class In(value: Expression, list: Seq[Expression]) extends Predicate `
  * `abstract class SubqueryExpression(`
  * `case class ListQuery(`
  * `case class Exists(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72992/
Test PASSed.

dilipbiswal commented:
cc @hvanhovell @gatorsmile @nsyca

gatorsmile commented:
cc @hvanhovell I did the internal review. It is ready for you to review it. Thanks!

nsyca commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -40,19 +42,179 @@ abstract class PlanExpression[T <: QueryPlan[_]] extends Expression {
 object SubqueryExpression {
+  /**
+   * Returns true when an expression contains an IN or EXISTS subquery and false otherwise.
+   */
```
The following definition is moved from `PredicateSubquery.hasPredicateSubquery` with the removing of the  PredicateSubquery in the first `case` pattern.

nsyca commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -40,19 +42,179 @@ abstract class PlanExpression[T <: QueryPlan[_]] extends Expression {
+  /**
+   * Returns whether there are any null-aware predicate subqueries inside Not. If not, we could
+   * turn the null-aware predicate into not-null-aware predicate.
+   */
```
The following definition is moved from `PredicateSubquery.hasNullAwarePredicateWithinNot` with a code change to reflect the new representation of IN subquery.

nsyca commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -40,19 +42,179 @@ abstract class PlanExpression[T <: QueryPlan[_]] extends Expression {
+  /**
+   * Returns true when an expression contains correlated predicates i.e outer references and
+   * returns false otherwise.
+   */
```
The following definition is moved from a nested definition of the same name in `ResovleSubquery.pullOuterCorrelatedPredicates` in `Analyzer.scala`.

nsyca commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -40,19 +42,179 @@ abstract class PlanExpression[T <: QueryPlan[_]] extends Expression {
+
+  /**
+   * Returns an expression after removing the OuterReference shell.
+   */
```
The following three definitions of `stripOuterReference(s)` is taken from a segment of code embedded in `ResolveSubquery.rewriteSubQuery` (`Analyzer.scala`).

nsyca commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -83,29 +95,150 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
+      outer: Seq[LogicalPlan]): (LogicalPlan, Seq[Expression]) = {
+    val predicateMap = scala.collection.mutable.Map.empty[LogicalPlan, Seq[Expression]]
+
+    /** Determine which correlated predicate references are missing from this plan. */
```
The following definition is moved unchanged from a nested definition of the same name in `ResolveSubquery.pullOutCorrelatedPredicates` (`Analyzer.scala`)

dilipbiswal commented on sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala:
```diff
@@ -622,7 +622,12 @@ class SubquerySuite extends QueryTest with SharedSQLContext {
   test("SPARK-15370: COUNT bug with attribute ref in subquery input and output ") {
     checkAnswer(
-      sql("select l.b, (select (r.c + count(*)) is null from r where l.a = r.c) from l"),
+      sql(
```
In this test, the query is not a valid query as its missing group by. It used to pass before as we implicitly added r.c into aggregate grouping as part of pulling up correlated predicates in analyzer.

dilipbiswal commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -83,29 +95,150 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
+  *
+  * TODO: Look to merge this rule with RewritePredicateSubquery.
+  */
+object PullupCorrelatedPredicates extends Rule[LogicalPlan] with PredicateHelper {
```
This function is mostly copied from Analyzer. The only difference is since validation is done by analyzer , we only process Filter, Project and Aggregate here.

dilipbiswal commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -365,17 +385,66 @@ object TypeCoercion {
 
+      // Handle type casting required between value expression and subquery output
+      // in IN subquery.
+      case i @ In(a, Seq(ListQuery(sub, children, exprId))) if !i.resolved =>
```
Plan example is available in [Code Changes](https://docs.google.com/document/d/18mqjhL9V1An-tNta7aVE13HkALRZ5GZ24AATA-Vqqf0/edit#) in section 3.

hvanhovell replied:
Maybe we should create a special InSubQuery expression. This looks like a lot of work for what we are doing here.

dilipbiswal commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -158,17 +160,7 @@ trait CheckAnalysis extends PredicateHelper {
-                  }
-
-                  cleanQuery(p.child)
+                case p: Project => cleanQuery(p.child)
```
Now the de-duplication of sub plan happens in optimizer when we actually pull up the correlated predicates. Thus the project case is simplified.

dilipbiswal commented:
@hvanhovell Hi Herman, was wondering if you had some time to look into this PR ? Please let me know your thoughts.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1398,42 +1399,46 @@ class Analyzer(
-        f.tupled(rewriteSubQuery(current, plans))
+        // Validate the outer reference and record the outer references as children of
+        // subquery expression.
+        f.tupled(current, checkAndGetOuterReferences(current))
```
NIT: Do we still need `f.tupled`?

dilipbiswal replied:
@hvanhovell We can remove tupled and just say f(current, ...) ?

hvanhovell replied:
I think we can. The reason I used `f.tupled` was because `rewriteSubQuery(..)` returned a tuple.

dilipbiswal replied:
@hvanhovell Got it. Thank you !!

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2512,3 +2522,67 @@ object ResolveCreateNamedStruct extends Rule[LogicalPlan] {
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    plan transform {
+      case f @ Filter(_, a: Aggregate) if f.resolved =>
```
This only works with aggregates that are already in the `Aggregate` operator, this seems like a regression. What does Hive do?

dilipbiswal replied:
@hvanhovell At the time of executing this rule, the aggregate expressions from the subquery plan are already pushed down to the Aggregate operator through ResolveAggregateFunctions. Here we are just updating the outer references in the subquery plan.

hvanhovell replied:
Ok, awesome.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -109,6 +109,26 @@ object TypeCoercion {
+    case (DateType, StringType) => Some(StringType)
+    case (StringType, TimestampType) => Some(StringType)
+    case (TimestampType, StringType) => Some(StringType)
+    case (TimestampType, DateType) => Some(StringType)
```
This seems weird. Is this also the current behavior?

dilipbiswal replied:
@hvanhovell Here is where i saw the code that handles the promotion between date and timestamp types.
https://github.com/dilipbiswal/spark/blob/33844bb8c6e72c35e12a7ea9f124ad3c6b8bd43c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L347:L360
Please let me know if i missed something here.

hvanhovell replied:
Shouldn't we factor that code out then? Now we have the same logic in two places.

dilipbiswal replied:
@hvanhovell Thanks!!. I had tried to do this before as well as this came up during the internal review. I have made another try. Please let me know what you think.

hvanhovell replied:
Ok please do this in a follow-up

hvanhovell replied:
O wait, you have changed it.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:
```diff
@@ -123,19 +123,36 @@ case class Not(child: Expression)
-    } else {
-      TypeCheckResult.TypeCheckSuccess
+    list match {
+      case ListQuery(sub, _, _) :: Nil =>
```
IIUC this is all done to get a better error message right?

dilipbiswal replied:
@hvanhovell Actually previously checkInputDataTypes() called from checkAnalysis never had to deal with **In subquery expressions** as it gets rewritten to PredicateSubquery. With the change in this PR we now see the original IN subquery expressions and thus we need to make sure the types are same between LHS and RHS of the IN subquery expression.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -40,19 +42,179 @@ abstract class PlanExpression[T <: QueryPlan[_]] extends Expression {
+   */
+  def hasNullAwarePredicateWithinNot(e: Expression): Boolean = {
+    e.find{ x =>
+      x.isInstanceOf[Not] && e.find {
```
I am pretty sure this is wrong. This will work on the following expression: `NOT(a) AND b IN(SELECT q FROM t)`

dilipbiswal replied:
@hvanhovell Actually I copied this function from existing PredicateSubquery.hasNullAwarePredicateWithinNot. I also found it a little odd that we are doing a e.find .. but was afraid to change this.

dilipbiswal replied:
@hvanhovell I actually had debugged this before and had completely forgotten about it. The case you mention actually works ok because of the way we invoke this functions.  Please see the caller at - [code](https://github.com/dilipbiswal/spark/blob/33844bb8c6e72c35e12a7ea9f124ad3c6b8bd43c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala#L197-L204)

dilipbiswal replied:
@hvanhovell I have refactored the code a bit and i hope, this function is bit clearer to follow. I have moved the code from its caller in checkAnalysis to this function. Please let me know what you think.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -40,19 +42,179 @@ abstract class PlanExpression[T <: QueryPlan[_]] extends Expression {
+  def getOuterReferences(conditions: Seq[Expression]): Seq[Expression] = {
+    val outerExpressions = scala.collection.mutable.ArrayBuffer.empty[Expression]
+    conditions foreach { expr =>
+      expr transformDown {
```
Use foreachDown?

dilipbiswal replied:
@hvanhovell Sure. I will change.

hvanhovell reviewed (changes requested):
I took an initial pass (not quite there yet). Most of it is in order. I left a few comments.

I will take another pass this week.

dilipbiswal commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1110,31 +1184,24 @@ class Analyzer(
      */
-    private def pullOutCorrelatedPredicates(sub: LogicalPlan): (LogicalPlan, Seq[Expression]) = {
-      val predicateMap = scala.collection.mutable.Map.empty[LogicalPlan, Seq[Expression]]
+    private def checkAndGetOuterReferences(sub: LogicalPlan): Seq[Expression] = {
```
@hvanhovell The code here validates the correlated references as well as collects them to rewrite the outer plan to record the outer references. You are suggesting to move the "check" portion to checkAnalysis ? I will change to use foreachUp. Thanks.

hvanhovell replied:
Ok, then leave as it is. I thought it was only doing validation.

dilipbiswal commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2332,6 +2337,11 @@ class Analyzer(
+      // the types between the value expression and list query expression of IN expression.
+      // We need to subject the subquery plan through ResolveTimeZone again to setup timezone
+      // information for time zone aware expressions.
+      case e: ListQuery => e.withNewPlan(ResolveTimeZone.apply(e.plan))
```
@hvanhovell Thank you.. I will change.

dilipbiswal commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -707,13 +709,85 @@ class Analyzer(
+     *    :     +- SubqueryAlias t1
+     *    :        +- Relation[c1#250,c2#251] parquet
+     *    +- Project [c1#299, c2#300]
+     *    +- Filter exists#272 [c1#299]
```
@hvanhovell Thanks.. will change.

dilipbiswal commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -707,13 +709,85 @@ class Analyzer(
+     *                 FROM t2
+     *                 WHERE t1.c1 = t2.c1)
+     * }}}
+      * Plan before resolveReference rule.
```
@hvanhovell I will work on a smaller example. I think we don't need the subquery expression in the left plan. That should cut down the plan size.

nsyca replied:
@dilipbiswal You probably are already on the right track that one of the subquery expression is not required. I guess either `T1 intersect T1 where exists (T2)` or a self-join scenario like `T1, T1 Tx where exists (T2 where <Q>.col = T2.col)` with `<Q>` be `T1` or `Tx` should also do.

dilipbiswal commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1110,31 +1184,24 @@ class Analyzer(
-    private def pullOutCorrelatedPredicates(sub: LogicalPlan): (LogicalPlan, Seq[Expression]) = {
-      val predicateMap = scala.collection.mutable.Map.empty[LogicalPlan, Seq[Expression]]
+    private def checkAndGetOuterReferences(sub: LogicalPlan): Seq[Expression] = {
+      val outerReferences = scala.collection.mutable.ArrayBuffer.empty[Seq[Expression]]
```
@hvanhovell Thank you. I will change to use ArrayBuffer[Expression] instead.

dilipbiswal committed:
Review comments

SparkQA commented:
**[Test build #73646 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73646/testReport)** for PR 16954 at commit [`886a744`](https://github.com/apache/spark/commit/886a744d8dbcb5108d569a6db688dcdb9dcbbe17).

dilipbiswal commented:
@hvanhovell Hello Herman, I have addressed all your comments except one where you are suggesting an optimization to failOnOuterReferenceInSubTree(). I am trying to see how i can improve   that code. In the meantime, i wanted to push the updates i have so far. Please go through when you have some time. Thanks  a lot.

SparkQA commented:
**[Test build #73646 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73646/testReport)** for PR 16954 at commit [`886a744`](https://github.com/apache/spark/commit/886a744d8dbcb5108d569a6db688dcdb9dcbbe17).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73646/
Test PASSed.

dilipbiswal committed:
review comments

SparkQA commented:
**[Test build #73712 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73712/testReport)** for PR 16954 at commit [`3b4bb90`](https://github.com/apache/spark/commit/3b4bb90deb34e6c1bb1671c76b66c83741937578).

SparkQA commented:
**[Test build #73712 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73712/testReport)** for PR 16954 at commit [`3b4bb90`](https://github.com/apache/spark/commit/3b4bb90deb34e6c1bb1671c76b66c83741937578).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73712/
Test PASSed.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -202,14 +194,9 @@ trait CheckAnalysis extends PredicateHelper {
-            }
+          case Filter(condition, _) if SubExprUtils.hasNullAwarePredicateWithinNot(condition) =>
+            failAnalysis(s"Null-aware predicate sub-queries cannot be used in nested" +
+              s" conditions: $condition")
```
Nit: `s"` -> `"` and move the space from the beginning of the second line to the end of the first line.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -365,17 +385,66 @@ object TypeCoercion {
+   * Handles type coercion for both IN expression with subquery and IN
+   * expressions without subquery.
+   * 1. In the first case, find the common type by comparing the left hand side
+   *    expression types against corresponding right hand side expression derived
```
`left hand side` -> `left hand side (LHS)`
`right hand side` -> `right hand side (RHS)`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -365,17 +385,66 @@ object TypeCoercion {
+      // Handle type casting required between value expression and subquery output
+      // in IN subquery.
+      case i @ In(a, Seq(ListQuery(sub, children, exprId))) if !i.resolved =>
+        // lhs is the value expression of IN subquery.
```
`lhs` -> `LHS`. Please correct all the similar cases in comments.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -365,17 +385,66 @@ object TypeCoercion {
+          findCommonTypeForBinaryComparison(l.dataType, r.dataType)
+        }
+
+        if (commonTypes.length == lhs.length) {
```
Please add a comment to explain what this condition means.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -365,17 +385,66 @@ object TypeCoercion {
+          // in a CreatedNamedStruct.
+          val newLhs = a match {
+            case cns: CreateNamedStruct =>
+              val nameValue = cns.nameExprs.zip(castedLhs).flatMap(pair => Seq(pair._1, pair._2))
```
Please use `case (name, value) =>` instead of `pair`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -54,20 +61,25 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
           val (joinCond, outerPlan) = rewriteExistentialExpr(conditions, p)
           Join(outerPlan, sub, LeftAnti, joinCond)
-        case (p, Not(PredicateSubquery(sub, conditions, true, _))) =>
+        case (p, In(e, Seq(l @ ListQuery(sub, conditions, _)))) =>
```
`case (p, In(e, Seq(ListQuery(sub, conditions, _))))`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -54,20 +61,25 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
+          val inConditions = getValueExpression(e).zip(sub.output).map(EqualTo.tupled)
+          val (joinCond, outerPlan) = rewriteExistentialExpr(inConditions ++ conditions, p)
+          Join(outerPlan, sub, LeftSemi, joinCond)
+        case (p, Not(In(e, Seq(l @ ListQuery(sub, conditions, _))))) =>
```
`case (p, Not(In(e, Seq(ListQuery(sub, conditions, _))))) =>`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -83,29 +95,150 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
       e transformUp {
-        case PredicateSubquery(sub, conditions, nullAware, _) =>
-          // TODO: support null-aware join
+        case Exists(sub, conditions, exprId) =>
```
`case Exists(sub, conditions, _)`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -83,29 +95,150 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
+        case Exists(sub, conditions, exprId) =>
           val exists = AttributeReference("exists", BooleanType, nullable = false)()
-          newPlan = Join(newPlan, sub, ExistenceJoin(exists), conditions.reduceLeftOption(And))
+          newPlan = Join(newPlan, sub,
```
`newPlan = Join(newPlan, sub, ExistenceJoin(exists), conditions.reduceLeftOption(And))`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -83,29 +95,150 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
+          val exists = AttributeReference("exists", BooleanType, nullable = false)()
+          val inConditions = getValueExpression(e).zip(sub.output).map(EqualTo.tupled)
+          newPlan = Join(newPlan, sub,
+            ExistenceJoin(exists), (inConditions ++ conditions).reduceLeftOption(And))
```
```
val newConditions = (inConditions ++ conditions).reduceLeftOption(And)
newPlan = Join(newPlan, sub, ExistenceJoin(exists), newConditions)
```

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -83,29 +95,150 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
+            ExistenceJoin(exists), conditions.reduceLeftOption(And))
           exists
-        }
+        case In(e, Seq(l@ ListQuery(sub, conditions, exprId))) =>
```
`case In(value, Seq(ListQuery(sub, conditions, _)))`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -83,29 +95,150 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
-        }
+        case In(e, Seq(l@ ListQuery(sub, conditions, exprId))) =>
+          val exists = AttributeReference("exists", BooleanType, nullable = false)()
+          val inConditions = getValueExpression(e).zip(sub.output).map(EqualTo.tupled)
```
`getValueExpression(e)` -> `getValueExpression(value)`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -83,29 +95,150 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
+    // In case of a collision, change the subquery plan's output to use
+    // different attribute by creating alias(s).
+    val baseConditions = predicateMap.values.flatten.toSeq
+    val (newplan: LogicalPlan, newcond: Seq[Expression]) = if (outer.nonEmpty) {
```
`newplan ` -> `newPlan `
`newcond` -> `newCond `

`val (newPlan, newCond) = if (outer.nonEmpty) {`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -83,29 +95,150 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
+
+  private def rewriteSubQueries(plan: LogicalPlan, outerPlans: Seq[LogicalPlan]): LogicalPlan = {
+    plan transformExpressions {
+      case s @ ScalarSubquery(sub, cond, exprId) if s.children.nonEmpty =>
```
`case ScalarSubquery(sub, children, exprId) if children.nonEmpty =>`

gatorsmile replied:
if you want, you also can use `cond` to replace `children`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -83,29 +95,150 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
+      case s @ ScalarSubquery(sub, cond, exprId) if s.children.nonEmpty =>
+        val (newPlan, newCond) = pullOutCorrelatedPredicates(sub, outerPlans)
+        ScalarSubquery(newPlan, newCond, exprId)
+      case e @ Exists(sub, cond, exprId) if e.children.nonEmpty =>
```
`case Exists(sub, children, exprId) if children.nonEmpty =>`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:
```diff
@@ -83,29 +95,150 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
+      case e @ Exists(sub, cond, exprId) if e.children.nonEmpty =>
+        val (newPlan, newCond) = pullOutCorrelatedPredicates(sub, outerPlans)
+        Exists(newPlan, newCond, exprId)
+      case l @ ListQuery(sub, cond, exprId) =>
```
`case ListQuery(sub, _, exprId) =>`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -79,57 +251,10 @@ case class ScalarSubquery(
-  def hasPredicateSubquery(e: Expression): Boolean = {
-    e.find {
-      case _: PredicateSubquery | _: ListQuery | _: Exists => true
+      case s: ScalarSubquery if s.children.nonEmpty => true
```
`case s: ScalarSubquery => s.children.nonEmpty`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -297,8 +284,11 @@ trait CheckAnalysis extends PredicateHelper {
+          case p if p.expressions.exists(SubqueryExpression.hasInOrExistsSubquery) =>
+            p match {
+              case _: Filter => // Ok
+              case other => failAnalysis(s"Predicate sub-queries can only be used in a Filter: $p")
```
`case _ => failAnalysis(`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1208,63 +1253,39 @@ class Analyzer(
-          p
         case p: Repartition =>
-          p
         case p: SubqueryAlias =>
```
`case _: BroadcastHint | _: Distinct | _: LeafNode | _: Repartition | _: SubqueryAlias =>`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1172,20 +1226,11 @@ class Analyzer(
-      val transformed = BooleanSimplification(sub) transformUp {
+      // Simplify the predicates before validating any unsupported correlation patterns
+      // in the plan.
+      BooleanSimplification(sub).foreachUp {
```
`BooleanSimplification ` is not needed after the changes of this PR

gatorsmile replied:
You should also remove the import of `BooleanSimplification`

dilipbiswal replied:
@gatorsmile Currently the code relies on this. I will need to investigate this and would like to address this in a followup.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1208,63 +1253,39 @@ class Analyzer(
           // Find all predicates with an outer reference.
-          val (correlated, local) = splitConjunctivePredicates(cond).partition(containsOuter)
+          val (correlated, local) =
+            splitConjunctivePredicates(cond).partition(SubExprUtils.containsOuter)
```
Let us `import org.apache.spark.sql.catalyst.expressions.SubExprUtils._` at the beginning of the rule of `ResolveSubquery`

Thus, no change is needed after the import.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1208,63 +1253,39 @@ class Analyzer(
-              predicateMap += child -> xs
-              child
-          }
+          outerReferences ++= SubExprUtils.getOuterReferences(correlated)
```
After the import, we also can clean up this to `outerReferences ++= getOuterReferences(correlated)`

gatorsmile commented:
@dilipbiswal I did one more pass. :)

dilipbiswal committed:
Review comments

SparkQA commented:
**[Test build #73746 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73746/testReport)** for PR 16954 at commit [`f2bf5d3`](https://github.com/apache/spark/commit/f2bf5d3187b2ab2ead527441bae083bb50420b11).

dilipbiswal commented:
@gatorsmile Thank you. I have addressed the comments.

dilipbiswal committed:
Review comment

SparkQA commented:
**[Test build #73748 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73748/testReport)** for PR 16954 at commit [`1191173`](https://github.com/apache/spark/commit/1191173cf42697c9543e0e4d8c8cbee385b0fe01).

SparkQA commented:
**[Test build #73750 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73750/testReport)** for PR 16954 at commit [`e43ac08`](https://github.com/apache/spark/commit/e43ac08c71c6634b4c1e2cb98bd278a7e36846dc).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73748/
Test PASSed.

SparkQA commented:
**[Test build #73750 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73750/testReport)** for PR 16954 at commit [`e43ac08`](https://github.com/apache/spark/commit/e43ac08c71c6634b4c1e2cb98bd278a7e36846dc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73750/
Test PASSed.

SparkQA commented:
**[Test build #73746 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73746/testReport)** for PR 16954 at commit [`f2bf5d3`](https://github.com/apache/spark/commit/f2bf5d3187b2ab2ead527441bae083bb50420b11).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73746/
Test PASSed.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -21,12 +21,13 @@ import scala.annotation.tailrec
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.expressions.aggregate._
 import org.apache.spark.sql.catalyst.expressions.objects.NewInstance
+import org.apache.spark.sql.catalyst.expressions.SubExprUtils._
```
Move it here? https://github.com/dilipbiswal/spark/blob/e43ac08c71c6634b4c1e2cb98bd278a7e36846dc/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L1148

dilipbiswal replied:
@gatorsmile There is another place in analyzer(updateOuterReferencesInSubquery) that also uses SubExprUtils. Thats why i moved it to the main imports .. What do you think ?

gatorsmile replied:
Ok

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1204,64 +1250,36 @@ class Analyzer(
           // Find all predicates with an outer reference.
-          val (correlated, local) = splitConjunctivePredicates(cond).partition(containsOuter)
+          val (correlated, local) =
+            splitConjunctivePredicates(cond).partition(containsOuter)
```
`val (correlated, local) = splitConjunctivePredicates(cond).partition(containsOuter)`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1395,42 +1367,45 @@ class Analyzer(
       } while (!current.resolved && !current.fastEquals(previous))
 
-      // Step 2: Pull out the predicates if the plan is resolved.
+      // Step 2: pull the outer references and record them as children of SubqueryExpression
```
`pull ` -> `Pull`. Also revert back the orignal msg `if the plan is resolved.`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1204,64 +1250,36 @@ class Analyzer(
 
         // Project cannot host any correlated expressions
         // but can be anywhere in a correlated subquery.
         case p @ Project(expressions, child) =>
```
`case p: Project =>`. The same issue in the following `Aggregate`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1204,64 +1250,36 @@ class Analyzer(
         // Category 3:
         // Filter is one of the two operators allowed to host correlated expressions.
         // The other operator is Join. Filter can be anywhere in a correlated subquery.
         case f @ Filter(cond, child) =>
```
`case f: Filter =>`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1317,62 +1327,24 @@ class Analyzer(
         // Generator with join=false is treated as Category 4.
         case p @ Generate(generator, true, _, _, _, _) =>
```
`case g: Generate if g.join =>`

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -200,14 +191,9 @@ trait CheckAnalysis extends PredicateHelper {
-            }
+          case Filter(condition, _) if hasNullAwarePredicateWithinNot(condition) =>
+            failAnalysis(s"Null-aware predicate sub-queries cannot be used in nested" +
+              " conditions: $condition")
```
```
failAnalysis("Null-aware predicate sub-queries cannot be used in nested " +
  s"conditions: $condition")
```

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -40,19 +43,194 @@ abstract class PlanExpression[T <: QueryPlan[_]] extends Expression {
+    e.transform {
+      case OuterReference(r) => r
+    }
+  }
```
```
def stripOuterReference(e: Expression): Expression = e.transform { case OuterReference(r) => r }
```

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -40,19 +43,194 @@ abstract class PlanExpression[T <: QueryPlan[_]] extends Expression {
+        case Nil =>
+        case xs =>
+          correlatedPredicates += xs
+      }
```
``` Scala
stripOuterReferences(corr) match {
  case Nil => // OK
  case xs => correlatedPredicates += xs
}
```

dilipbiswal committed:
code review

SparkQA commented:
**[Test build #73789 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73789/testReport)** for PR 16954 at commit [`1081c9f`](https://github.com/apache/spark/commit/1081c9fda7d19971941d677016f2f9078085abe8).

SparkQA commented:
**[Test build #73789 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73789/testReport)** for PR 16954 at commit [`1081c9f`](https://github.com/apache/spark/commit/1081c9fda7d19971941d677016f2f9078085abe8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73789/
Test PASSed.

gatorsmile commented:
See the comment https://github.com/apache/spark/pull/16954#discussion_r104287344

dilipbiswal committed:
Review comments

SparkQA commented:
**[Test build #73915 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73915/testReport)** for PR 16954 at commit [`7178719`](https://github.com/apache/spark/commit/7178719aaae961a3b5b38132d09a0d4d91ade692).

SparkQA commented:
**[Test build #73915 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73915/testReport)** for PR 16954 at commit [`7178719`](https://github.com/apache/spark/commit/7178719aaae961a3b5b38132d09a0d4d91ade692).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73915/
Test PASSed.

dilipbiswal committed:
rebase

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -365,17 +368,73 @@ object TypeCoercion {
+
+        // RHS is the subquery output.
+        val rhs = sub.output
+        require(lhs.length == rhs.length)
```
Please check this in the guard of the rule. This currently throws a very hard to understand error.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -365,17 +368,73 @@ object TypeCoercion {
+        require(lhs.length == rhs.length)
+
+        val commonTypes = lhs.zip(rhs).flatMap { case (l, r) =>
+          findCommonTypeForBinaryComparison(l.dataType, r.dataType) match {
```
Use `orElse`?

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -365,17 +368,73 @@ object TypeCoercion {
+
+          // Before constructing the In expression, wrap the multi values in LHS
+          // in a CreatedNamedStruct.
+          val newLhs = a match {
```
I think it is better if you cast on the `castedLhs` here:
```scala
val newLhs = castedLhs match {
  case Seq(lhs) => lhs
  case _ =>
    CreateNamedStruct(cns.nameExprs.zip(castedLhs).flatMap {
      case (name, value) => Seq(name, value)
    })
}
```

dilipbiswal replied:
@hvanhovell Can you please double check the above code ? I am unable to compile the code as cns is not defined. In this block, i was looking at the original left hand side expression to see if its a named struct and if so i construct a named struct back. castedLhs would always match a Seq(lhs) , no ? Please let me know if i am missing something here.

hvanhovell replied:
My bad, I never compile these snippets. You have a point there. We could just use CreateStruct (since we really don't care about the name). So that would look something like this (again not compiled):
```scala
val newLhs = castedLhs match {
  case Seq(lhs) => lhs
  case _ => CreateStruct(castedLhs)
}
```

dilipbiswal replied:
@hvanhovell Thanks a lot. You are right, we don't care about the names. This looks much better.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:
```diff
@@ -123,19 +123,36 @@ case class Not(child: Expression)
+        val isTypeMismatched = valExprs.zip(sub.output).exists {
+          case (l, r) => l.dataType != r.dataType
+        }
+        if (isTypeMismatched) {
```
It is better to display the offending columns. Now the user has to find the offending columns by herself/himself.

dilipbiswal replied:
@hvanhovell The new error message looks like following. Does this look okay to you ?

```
Error in query: cannot resolve '(named_struct('c1', at1.`c1`, 'c2', at1.`c2`) IN (listquery()))' due to data type mismatch: 
The data type of one or more elements in the left hand side of an IN subquery
is not compatible with the data type of the output of the subquery
Mismatched columns:
[(at1.`c1`:decimal(10,0), at2.`c1`:timestamp), (at1.`c2`:timestamp, at2.`c2`:decimal(10,0))]
Left side:
[decimal(10,0), timestamp].
Right side:
[timestamp, decimal(10,0)].
```

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:
```diff
@@ -123,19 +123,36 @@ case class Not(child: Expression)
+          TypeCheckResult.TypeCheckFailure(
+            s"""
+              |The data type of one or more elements in the LHS of an IN subquery
+              |[${valExprs.map(_.dataType).mkString(", ")}]
```
Use catalog strings please.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -40,19 +43,189 @@ abstract class PlanExpression[T <: QueryPlan[_]] extends Expression {
+      case _: Exists | Not(_: Exists) | In(_, Seq(_: ListQuery)) | Not(In(_, Seq(_: ListQuery))) =>
+        false
+      case e => e.find { x =>
+        x.isInstanceOf[Not] && e.find {
```
This still looks fishy to me. It is also in the original code base, so we don't have to fix this now. Can you open a JIRA to track this?

dilipbiswal replied:
@hvanhovell Sure. will do

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -40,19 +43,189 @@ abstract class PlanExpression[T <: QueryPlan[_]] extends Expression {
+    val outerExpressions = ArrayBuffer.empty[Expression]
+    conditions foreach { expr =>
+      expr transformDown {
+        case a: AggregateExpression if containsOuter(a) =>
```
Shouldn't we check if the aggregate contains inner references? You could use collect leaves here: `a.collectLeaves.forall(_.isInstanceOf[OuterReference])`

dilipbiswal replied:
@hvanhovell Sounds good Herman. I will change the guard to look for "all outer references" now. I will do a small followup to raise an appropriate exception if there are mixture of outer and inner references within a AggregateExpression as i have to write new tests for it. Hope thats ok with you.

hvanhovell replied:
Yeah just address this in a follow-up

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -40,19 +43,189 @@ abstract class PlanExpression[T <: QueryPlan[_]] extends Expression {
+    val conditions = plan.collect { case Filter(cond, _) => cond }
+
+    // Collect all the expressions that have outer references.
+    conditions foreach { e =>
```
Just use flatMap here? Instead of foreach?

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1204,80 +1250,44 @@ class Analyzer(
-              predicateMap += child -> xs
-              child
-          }
+          outerReferences ++= getOuterReferences(correlated)
```
`getOuterReferences` is kind of magical in the sense that it also isolates the aggregate expression. Could you add a single line of comment to highlight this?

dilipbiswal replied:
@hvanhovell Will do.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:
```diff
@@ -40,19 +43,189 @@ abstract class PlanExpression[T <: QueryPlan[_]] extends Expression {
   }
 }
 
+object SubExprUtils extends PredicateHelper {
```
Why make create two objects `SubqueryExpression` and `SubExprUtils`? Are they that different?

dilipbiswal replied:
@hvanhovell Actually @nsyca and i had discussed a bit about this. SubqueryExpression object has methods that operates strictly in the context of SubqueryExpression where as the utils has methods that mostly deals with OuterReferences.. So they can operate on the subquery plans referred to from SubqueryExpression or may be in the future if we support  queries of the form.
```
select * from t1 left outer join (select * from t2 where t2.c1 = t1.c1) on …
```
If you think we should merge this two then let me know and i will do it :-).

hvanhovell replied:
Ok, so the the `SubExprUtils` should be moved to the `OuterReference` companion :)...

Let's leave it for now.

hvanhovell commented:
@dilipbiswal can you update. I left a few minor comments, but it is good to go anyway (pending the update).

dilipbiswal commented:
@hvanhovell I am in the process of incorporating the comments. I will send an update soon.

dilipbiswal committed:
Code review

SparkQA commented:
**[Test build #74489 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74489/testReport)** for PR 16954 at commit [`19cdbb0`](https://github.com/apache/spark/commit/19cdbb040ccf2e74e1271ca33e6842607c1e0760).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74489/
Test FAILed.

dilipbiswal commented:
retest this please

SparkQA commented:
**[Test build #74495 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74495/testReport)** for PR 16954 at commit [`19cdbb0`](https://github.com/apache/spark/commit/19cdbb040ccf2e74e1271ca33e6842607c1e0760).

SparkQA commented:
**[Test build #74495 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74495/testReport)** for PR 16954 at commit [`19cdbb0`](https://github.com/apache/spark/commit/19cdbb040ccf2e74e1271ca33e6842607c1e0760).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74495/
Test PASSed.

hvanhovell commented:
LGTM. Merging to master. Thanks for the hard work, and your patience!

dilipbiswal commented:
@hvanhovell @gatorsmile @nsyca Thank you very much !!
